### PR TITLE
client: check OSD caps before read/write

### PR DESCRIPTION
**Pull request by ukernel**
Signed-off-by: Yan, Zheng <zyan@redhat.com>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -10740,6 +10745,76 @@ bool Client::is_quota_bytes_approaching(Inode *in)
...
+    objecter->mutate(oid, OSDMap::file_to_object_locator(in->layout), wr_op,
+		     in->snaprealm->get_snap_context(), ceph_clock_now(cct), 0,
+		     new C_SafeCond(&lock, &cond, &done[1], &wr_ret), NULL);
+
+    client_lock.Unlock();
+    lock.Lock();
+    while (!done[0] || !done[1])
+      cond.Wait(lock);
+    lock.Unlock();
+    client_lock.Lock();
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: This looks racy to me — certainly we could spin off somebody else checking if we can access the same pool, at minimum, but I'm concerned something more dangerous might be possible too. Like maybe getting caps revoked while we're in the process of testing? :/

**Comment by gregsfortytwo** [comment]
So this looks about like @liewegas and I have discussed in the past, but we've never been satisfied with the solution so we haven't tried to implement it.

As we discussed in standup, I'm particularly concerned about this failing once we add more sophisticated permissions checking to CephFS. The existing options we could incorporate are namespace permissions (which would require client changes to handle properly) and prefix permissions (where we allow clients to access only objects with a certain prefix — we could do this by allocating specific inode ranges to each client on the MDS or something, and this would not require any client changes to function correctly).

But we could also do a whole lot of other things in the future which we haven't thought of yet. For instance, I'm not sure how we could extend this to something where we submit class operations to the file data objects and they decide whether to accept or reject us based on some other criteria.

On the other hand, maybe we are okay with making people upgrade if they want better permissions models to be handled correctly. Sage, any thoughts?

File: src/client/Client.cc (hunk 1 above)
Comment (liewegas) [inline src/client/Client.cc]: It look sto me like we might be multiple threads checking the same thing, but nothing bad will happen.

**Comment by liewegas** [comment]
Yeah, eventually we'll need to address all that, but I think we can wait until we get there to make something more complicated.  In the meantime, this is pretty simple and makes life better... *shrug*

File: src/client/Client.cc (hunk 1 above)
Comment (ukernel) [inline src/client/Client.cc]: On Sat, Apr 25, 2015 at 5:20 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/client/Client.cc
> <https://github.com/ceph/ceph/pull/4457#discussion_r29086863>:
>
> > +		     in->snaprealm->get_snap_context(), ceph_clock_now(cct), 0,
> > +		     new C_SafeCond(&lock, &cond, &done[0], &rd_ret), NULL);
> > +
> > +    ObjectOperation wr_op;
> > +    wr_op.create(true);
> > +
> > +    objecter->mutate(oid, OSDMap::file_to_object_locator(in->layout), wr_op,
> > +		     in->snaprealm->get_snap_context(), ceph_clock_now(cct), 0,
> > +		     new C_SafeCond(&lock, &cond, &done[1], &wr_ret), NULL);
> > +
> > +    client_lock.Unlock();
> > +    lock.Lock();
> > +    while (!done[0] || !done[1])
> > +      cond.Wait(lock);
> > +    lock.Unlock();
> > +    client_lock.Lock();
>
> This looks racy to me -- certainly we could spin off somebody else checking
> if we can access the same pool, at minimum, but I'm concerned something
> more dangerous might be possible too. Like maybe getting caps revoked while
> we're in the process of testing? :/
>
this check happens before we get caps reference, I can't see any caps
revoke race.

**Comment by ukernel** [comment]
updated with following changes:
---
avoid concurrent checking
and a config option to disable the check

**Comment by gregsfortytwo** [comment]
Okay, this looks good to me, but we need a test to vaildate that it's behaving properly. Should be pretty easy to stick a two-client test in the qa suite that denies access for one client, allows it for the other, and shows that they behave as expected.

**Comment by ukernel** [comment]
https://github.com/ceph/ceph-qa-suite/pull/421

**Comment by ukernel** [comment]
http://pulpito.ceph.com/compare/?branch=wip-client-osd-caps&suite=fs

only failure was caused by "cluster [ERR] bad backtrace on dir ino 10000000003" in cluster log

**Pull request merged**